src/workerd/server/server.c++

```diff
@@ -497,6 +497,51 @@ private:
   kj::Maybe<kj::Own<kj::NetworkAddress>> addr;
 };
 
+class Server::ExternalTcpService final: public Service, private WorkerInterface {
+public:
+  ExternalTcpService(kj::Own<kj::NetworkAddress> addrParam)
+    : addr(kj::mv(addrParam)) {}
+
+  kj::Own<WorkerInterface> startRequest(IoChannelFactory::SubrequestMetadata metadata) override {
+    return { this, kj::NullDisposer::instance };
+  }
+
+  bool hasHandler(kj::StringPtr handlerName) override {
+    return handlerName == "fetch"_kj || handlerName == "connect"_kj;
+  }
+
+private:
+  kj::Own<kj::NetworkAddress> addr;
+  
+  kj::Promise<void> request(
+      kj::HttpMethod method, kj::StringPtr url, const kj::HttpHeaders& headers,
+      kj::AsyncInputStream& requestBody, kj::HttpService::Response& response) override {
+    throwUnsupported();
+  }
+
+  kj::Promise<void> connect(
+      kj::StringPtr host, const kj::HttpHeaders& headers, kj::AsyncIoStream& connection,
+      ConnectResponse& tunnel, kj::HttpConnectSettings settings) override {
+    connection = *(co_await addr->connect());
```

Comment:
Huh, I'm surprised this works without actually calling `accept()` on the `ConnectResponse`. Is that just its default behavior if you ignore it?

For that matter, I'm a little surprised/impressed that just assigning to an abstract class reference like this works.

But please just ignore this confusion if you've already confirmed it all works.

Reply:
Yeah this code definitely doesn't do the right thing.

I would not have expected this line to compile. If it does compile, it must be a no-op, because `AsyncIoStream`, being an abstract interface, has no fields, and thus the auto-generated assignment operator has nothing to do. (But if the compiler actually lets you perform default assignment on an abstract type without even a warning, I'm surprised...)

Reply:
Ah my bad, I misunderstood how this worked under the hood. I re-implemented following `HttpServiceAdapter`'s connect implementation